src/strval.rs

```diff
@@ -46,58 +46,254 @@ impl From<()> for StrValError {
         }
-        "u64" => ScVal::Object(Some(ScObject::U64(val.parse()?))),
-        "sym" => ScVal::Symbol(
-            val.as_bytes()
+
+        // For all others we just use the json parser
+        _ => serde_json::from_str(s)
+            .map_err(StrValError::Serde)
+            .and_then(|raw| from_json(&raw, t))?,
+    };
+    Ok(val)
+}
+
+#[allow(clippy::too_many_lines)]
+pub fn from_json(v: &Value, t: &ScSpecTypeDef) -> Result<ScVal, StrValError> {
+    let val: ScVal = match (t, v) {
+        // Boolean parsing
+        (ScSpecTypeDef::Bool, Value::Bool(true)) => ScVal::Static(ScStatic::True),
+        (ScSpecTypeDef::Bool, Value::Bool(false)) => ScVal::Static(ScStatic::False),
+
+        // Vec parsing
+        (ScSpecTypeDef::Vec(elem), Value::Array(raw)) => {
+            let ScSpecTypeVec { element_type } = &**elem;
+            let parsed: Result<Vec<ScVal>, StrValError> = raw
+                .iter()
+                .map(|item| -> Result<ScVal, StrValError> { from_json(item, element_type) })
+                .collect();
+            let converted: ScVec = parsed?.try_into().map_err(StrValError::Xdr)?;
+            ScVal::Object(Some(ScObject::Vec(converted)))
+        }
+
+        // Number parsing
+        (ScSpecTypeDef::BigInt, Value::String(s)) => from_string(s, &ScSpecTypeDef::BigInt)?,
+        (ScSpecTypeDef::BigInt, Value::Number(n)) => {
+            from_json(&Value::String(format!("{}", n)), &ScSpecTypeDef::BigInt)?
+        }
+        (ScSpecTypeDef::I32, Value::Number(n)) => ScVal::I32(
+            n.as_i64()
+                .ok_or(StrValError::InvalidValue)?
+                .try_into()
+                .map_err(|_| StrValError::InvalidValue)?,
+        ),
+        (ScSpecTypeDef::I64, Value::Number(n)) => ScVal::Object(Some(ScObject::I64(
+            n.as_i64().ok_or(StrValError::InvalidValue)?,
+        ))),
+        (ScSpecTypeDef::U32, Value::Number(n)) => ScVal::U32(
+            n.as_u64()
+                .ok_or(StrValError::InvalidValue)?
+                .try_into()
+                .map_err(|_| StrValError::InvalidValue)?,
+        ),
+        (ScSpecTypeDef::U64, Value::Number(n)) => {
+            ScVal::U63(n.as_i64().ok_or(StrValError::InvalidValue)?)
+        }
+
+        // Map parsing
+        (ScSpecTypeDef::Map(map), Value::Object(raw)) => {
+            let ScSpecTypeMap {
+                key_type,
+                value_type,
+            } = &**map;
+            // TODO: What do we do if the expected key_type is not a string or symbol?
+            let parsed: Result<Vec<ScMapEntry>, StrValError> = raw
+                .iter()
+                .map(|(k, v)| -> Result<ScMapEntry, StrValError> {
+                    let key = from_string(k, key_type)?;
+                    let val = from_json(v, value_type)?;
+                    Ok(ScMapEntry { key, val })
+                })
+                .collect();
+            let converted: ScMap = parsed?.try_into().map_err(StrValError::Xdr)?;
```

Comment:
I recommend using `ScMap::sorted_from` here. I don't know if it's needed, but this has bit me on two separate occasions now (see https://github.com/stellar/rs-stellar-xdr/issues/98 for context). Better safe than sorry on this, because it is extremely difficult to debug errors related to that issue.

If this kind of parsing is later used to facilitate signing then this is even more likely to be important, because that is where I have encountered the error.